cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -516,6 +684,125 @@ class HashJoinNode : public ExecNode {
 
   const char* kind_name() const override { return "HashJoinNode"; }
 
+  Status OnBuildSideBatch(size_t thread_index, ExecBatch batch) {
+    std::lock_guard<std::mutex> guard(build_side_mutex_);
+    accumulators_[1].InsertBatch(std::move(batch));
+    return Status::OK();
+  }
+
+  Status OnBuildSideFinished(size_t thread_index) {
+    return pushdown_context_.BuildBloomFilter(
+        thread_index, std::move(accumulators_[1]),
+        [this](size_t thread_index, AccumulationQueue batches) {
+          return OnBloomFilterFinished(thread_index, std::move(batches));
+        });
+  }
+
+  Status OnBloomFilterFinished(size_t thread_index, AccumulationQueue batches) {
+    RETURN_NOT_OK(pushdown_context_.PushBloomFilter());
+    return impl_->BuildHashTable(
+        thread_index, std::move(batches),
+        [this](size_t thread_index) { return OnHashTableFinished(thread_index); });
+  }
+
+  Status OnHashTableFinished(size_t thread_index) {
+    bool should_probe;
+    {
+      std::lock_guard<std::mutex> guard(probe_side_mutex_);
+      should_probe = queued_batches_filtered_ && !hash_table_ready_;
+      hash_table_ready_ = true;
+    }
+    if (should_probe) {
+      return ProbeQueuedBatches(thread_index);
+    }
+    return Status::OK();
+  }
+
+  Status OnProbeSideBatch(size_t thread_index, ExecBatch batch) {
+    bool should_filter;
+    bool should_accum;
+    {
+      std::lock_guard<std::mutex> guard(probe_side_mutex_);
+      should_filter = bloom_filters_ready_;
+      should_accum = !should_filter || !hash_table_ready_;
+      if (should_accum) {
```

Comment:
I'm not 100% sure I'm following this but it seems like `should_accum` will be true if the bloom filters are ready but the hash table is not ready.  Couldn't we run filter single batch and then accumulate in that case?

Reply:
Yep, I think I've simplified this section to be correct.